### PR TITLE
PyTorch: update ecosystem

### PR DESCRIPTION
**Pull request by adamjstewart**
* https://github.com/pytorch/pytorch/releases/tag/v2.4.0
* https://github.com/pytorch/audio/releases/tag/v2.4.0
* https://github.com/pytorch/data/releases/tag/v0.8.0
* https://github.com/pytorch/vision/releases/tag/v0.19.0

### CODE REVIEW

**Comment by adamjstewart** [comment]
@spackbot run pipeline

**Comment by adamjstewart** [comment]
@spackbot run pipeline

**Comment by blapie** [comment]
Thanks Adam! From a SLEEF point of view this lgtm.

File: var/spack/repos/builtin/packages/py-torch/package.py [hunk 1]
```diff
@@ -239,7 +238,8 @@ class PyTorch(PythonPackage, CudaPackage, ROCmPackage):
         depends_on("py-pybind11@2.10.0", when="@1.13:1")
         depends_on("py-pybind11@2.6.2", when="@1.8:1.12")
         depends_on("py-pybind11@2.3.0", when="@:1.7")
-    depends_on("sleef@3.5.1_2020-12-22", when="@1.8:")
+    depends_on("sleef@3.6.0_2024-03-20", when="@2.4:")
```
Comment (blapie) [inline var/spack/repos/builtin/packages/py-torch/package.py:238]: Is there a particular reason for not mapping 2.4.0 to a newer version of SLEEF?

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-torch/package.py:238]: PyTorch is picky. All dependencies are vendored at specific commits. Historically, every time I've tried to relax these (even trying the stable release before or after), I've encountered a slew of build errors. Most dependencies don't even have stable releases.

I kinda gave up and stopped trying to convince the PyTorch devs to follow best practices. Anyone willing to 1) test ranges of versions for dependencies, 2) update PyTorch to use stable versions, 3) unvendor more dependencies, etc. is very welcome to try.

Comment (blapie) [reply var/spack/repos/builtin/packages/py-torch/package.py:238]: I understand, thanks for explaining!

**Review by blapie — APPROVED** [review]

**Comment by adamjstewart** [comment]
@spackbot run pipeline

**Comment by adamjstewart** [comment]
Don't think it will be possible to get the build to succeed without https://github.com/spack/spack/pull/45130

**Comment by haampie** [comment]
To get this in earlier, turn `USE_FLASH_ATTENTION` off (it defaults to ON on non-windows when cuda or rocm is used).

Or create a variant and add a requirement in the gitlab stacks that turns it off.

**Comment by adamjstewart** [comment]
Variant is an absolute nightmare, going to just disable it until it's no longer broken.

File: var/spack/repos/builtin/packages/py-torchvision/package.py [hunk 2]
```diff
@@ -118,6 +121,7 @@ class PyTorchvision(PythonPackage):
     depends_on("cuda", when="+nvjpeg")
     depends_on("ffmpeg@3.1:", when="+ffmpeg")
     depends_on("cuda", when="+video_codec")
+    depends_on("giflib", when="+gif")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-torchvision/package.py]: TODO: I'm not even sure if it's possible to build against external giflib installations.

**Review by alalazo — APPROVED** [review]

File: var/spack/repos/builtin/packages/py-torch/package.py [hunk 3]
```diff
@@ -559,6 +561,11 @@ def enable_or_disable(variant, keyword="USE", var=None):
             env.set("CUDNN_INCLUDE_DIR", self.spec["cudnn"].prefix.include)
             env.set("CUDNN_LIBRARY", self.spec["cudnn"].libs[0])
 
+        # Flash attention has very high memory requirements that may cause the build to fail
+        # https://github.com/pytorch/pytorch/issues/111526
+        # https://github.com/pytorch/pytorch/issues/124018
+        env.set("USE_FLASH_ATTENTION", "OFF")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-torch/package.py:567]: For the record, this results in the following warning message during use:
```
UserWarning: Torch was not compiled with flash attention.
```
Would love to find a better workaround.

**Pull request merged**